### DISCUSSION TITLE
Auto-generating API docs from YAML front matter

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3363] [Show and tell]
I've spent the better half of a day making some custom template files which I use to auto-generate page content for my API docs.
The docs in question aren't a REST API and more what you may get with libraries like Python (In my case is it Java).

I first had these pages made manually, which was a huge pain, especially since I was using markdown and am kind of a nitpicky guy (i.e. want to have tables look beautiful, so I pad their with to make them fit), so I decided to work on this stuff to actually generate the content automatically by pulling data from a page's YAML frontmatter.

I simply couldn't think of a more elegant solution, but for what it is does it work really well and *"easy"* (Depending on how easy you find it to edit YAML stuff...

I'll give some quick examples here to look at, with a link, the raw page content and a preview image (Caching may cause some visual differences).:

<details><summary>ProfileEntry</summary>
<p>

https://docs.andre601.ch/AdvancedServerList/api/reference/api/profiles/profileentry/

```markdown
---
template: api-doc.html

constructors:
  - name: 'ProfileEntry(List<String>, List<String>, String, String, NullBool, NullBool, Integer)'
    description: |
      Creates a new instance of a ProfileEntry with the given values.<br>
      It's recommended to use the <a href="./builder/"><code>Builder</code> class</a> for a more convenient configuration of the Settings.
    parameters:
      - name: 'motd'
        description: 'The MOTD to use.'
        attributes:
          - notnull
      - name: 'players'
        description: 'The players (Lines) to show in the hover.'
        attributes:
          - notnull
      - name: 'playerCountText'
        description: 'The text to display instead of the player count.'
        attributes:
          - nullable
      - name: 'favicon'
        description: 'The favicon to use.'
        attributes:
          - nullable
      - name: 'hidePlayersEnabled'
        description: 'Whether player count should be hidden.'
        attributes:
          - notnull
      - name: 'extraPlayersEnabled'
        description: 'Whether the extra players option should be enabled.'
        attributes:
          - notnull
      - name: 'extraPlayersCount'
        description: 'The number to add to the online players for the max players number.'
        attributes:
          - nullable
    seealso:
      - name: 'ProfileEntry.Builder'
        link: './builder/'

classes:
  - name: 'Builder'
    description: 'Builder class to create a new ProfileEntry instance.'
    type: 'static class'
    link: 'builder/'

methods:
  - name: 'empty()'
    description: |
      Creates an "empty" PlayerEntry with the following values set:<br>
      <ul>
        <li><a href="#motd()"><code>motd</code></a>: Empty List</li>
        <li><a href="#players()"><code>players</code></a>: Empty List</li>
        <li><a href="#playercounttext()"><code>playerCountText</code></a>: Empty String</li>
        <li><a href="#favicon()"><code>favicon</code></a>: Empty String</li>
        <li><a href="#hideplayersenabled()"><code>hidePlayersEnabled</code></a>: <a href="../../objects/nullbool/"><code>NullBool.NOT_SET</code></a></li>
        <li><a href="#extraplayersenabled()"><code>extraPlayersEnabled</code></a>: <a href="../../objects/nullbool/"><code>NullBool.NOT_SET</code></a></li>
        <li><a href="#extraplayerscount()"><code>extraPlayersCount</code></a>: <code>null</code></li>
      </ul>
    returns: 'New ProfileEntry instance with empty/null values defined'
    attributes:
      - static
    type:
      name: 'ProfileEntry'
      type: 'object'
  - name: 'copy()'
    description: |
      Creates a copy of this ProfileEntry instance.<br>
      <br>
      This is simply a convenience method to not have to call <a href="#builder()"><code>builder()</code></a><code>.</code><a href="./builder/#build()"><code>build()</code></a>.<br>
      If you want to modify the ProfileEntry <a href="#builder()"><code>builder()</code></a> is preferred.
    returns: 'A copy of this ProfileEntry instance.'
    seealso:
      - name: 'builder()'
        link: '#builder()'
    type:
      name: 'ProfileEntry'
      type: 'object'
  - name: 'builder()'
    description: |
      Creates a <a href="./builder/"><code>Builder</code> instance</a> with the values from this ProfileEntry set.<br>
      Use this method if you would like to modify the ProfileEntry.
    returns: 'A new <a href="./builder/"><code>Builder</code> instance</a> with the values of this ProfileEntry set.'
    type:
      name: 'Builder'
      type: 'object'
      link: './builder/'
  - name: 'motd()'
    description: 'Gets the currently set MOTD of this ProfileEntry.'
    returns: 'The current MOTD used by this ProfileEntry.'
    type:
      name: 'List<String>'
      type: 'object'
  - name: 'players()'
    description: 'Gets the currently set players of this ProfileEntry.'
    returns: 'The current list of players used by this ProfileEntry.'
    type:
      name: 'List<String>'
      type: 'object'
  - name: 'playerCountText()'
    description: 'Gets the currently set player count text of this ProfileEntry.'
    returns: 'The current player count text used by this ProfileEntry.'
    type:
      name: 'String'
      type: 'object'
  - name: 'favicon()'
    description: |
      Gets the currently set favicon of this ProfileEntry.<br>
      Note that the favicon usually is and supports one of the following options:
      <ul>
        <li>URL to a valid PNG file</li>
        <li>File name (with <code>.png</code> extension) matching a file saved in the favicons folder of AdvancedServerList</li>
        <li><code>${player uuid}</code> to display the avatar of the player</li>
      </ul>
    returns: 'The current favicon used by this ProfileEntry.'
    type:
      name: 'String'
      type: 'object'
  - name: 'hidePlayersEnabled()'
    description: |
      Whether the player count should be hidden or not in this ProfileEntry.<br>
      To get the actual boolean value, append <a href="../../objects/nullbool/#getordefault()"><code>getOrDefault(boolean)</code></a>.
    returns: 'Whether the player count should be hidden or not in this ProfileEntry.'
    type:
      name: 'NullBool'
      type: 'object'
      link: '../../objects/nullbool/'
  - name: 'extraPlayersEnabled()'
    description: |
      Whether the extra players feature should be used or not in this ProfileEntry.<br>
      To get the actual boolean value, append <a href="../../objects/nullbool/#getordefault()"><code>getOrDefault(boolean)</code></a>.
    returns: 'Whether the extra players feature should be used or not in this ProfileEntry.'
    type:
      name: 'NullBool'
      type: 'object'
      link: '../../objects/nullbool/'
  - name: 'extraPlayersCount()'
    description: 'Gets the currently set number of extra players to use by this ProfileEntry. May be <code>null</code>.'
    returns: 'Possibly-null integer number of extra players used by this ProfileEntry.'
    attributes:
      - nullable
    type:
      name: 'Integer'
      type: 'object'
  - name: 'isInvalid()'
    description: |
      Whether this ProfileEntry is considered invalid or not.<br>
      The ProfileEntry is considered invalid if all the following is true:
      <ul>
        <li><a href="#motd()"><code>motd()</code></a> is empty.</li>
        <li><a href="#players()"><code>players()</code></a> is empty.</li>
        <li><a href="#playercounttext()"><code>playerCountText()</code></a> is null/empty <b>and</b> <a href="#hideplayersenabled()"><code>hidePlayersEnabled()</code></a> is <code>false</code>.</li>
        <li><a href="#favicon()"><code>favicon()</code></a> is null/empty.</li>
      </ul>
      As long as one of the above is <b>not</b> true is this ProfileEntry considered valid and <code>false</code> will be returned.
    returns: 'Whether this ProfileEntry is invalid or not.'
    type:
      name: 'boolean'
      type: 'primitive'
---

# <api__record></api__record> ProfileEntry

This record represents the content found in a server list profile YAML file.  
The content may come from either the "profiles" list, the options in the file itself (global options) or a mix of both.

This class is immutable. Use [`builder()`](#builder()) to get a [`Builder` instance](builder.md) with the values of this class added.
```

![grafik](https://github.com/mkdocs/mkdocs/assets/11576465/db0b8948-db0a-41f4-90df-332e81c5631a)

</p>
</details> 

<details><summary>PlaceholderProvider</summary>
<p>

https://docs.andre601.ch/AdvancedServerList/api/reference/api/placeholderprovider/

```markdown
---
template: api-doc.html

constructors:
  - name: 'PlaceholderProvider(String)'
    description: 'Constructor used to set the identifier for the class extending the PlaceholderProvider class itself.'
    parameters:
      - name: 'identifier'
        description: 'The identifier to use for the placeholder. Cannot be empty.'
        attributes:
          - notnull

methods:
  - name: 'parsePlaceholder(String, GenericPlayer, GenericServer)'
    description: |
      Method called by AdvancedServerList's StringReplacer class to replace any appearances of <code>${&lt;identifier&gt; &lt;placeholder&gt;}</code> with whatever value a matching PlaceholderProvider may return.<br>
      <br>
      Returning <code>null</code> will be treated as an invalid placeholder by the plugin, making it return the placeholder as-is without any changes.
    parameters:
      - name: 'placeholder'
        description: 'The part of the Placeholder that comes after the identifier and before the closing curly bracket.'
      - name: 'player'
        description: 'The <a href="./../objects/genericplayer/"><code>GenericPlayer</code> instance</a> used.'
      - name: 'server'
        description: 'The <a href="./../objects/genericserver/"><code>GenericServer</code> instance</a> used.'
    returns: 'Possibly-parsed or possibly-null String to replace the placeholder with.'
    attributes:
      - nullable
    type:
      name: 'String'
      type: 'object'
  - name: 'getIdentifier()'
    description: 'Returns the identifier used by this PlaceholderProvider instance.'
    returns: 'String representing the identifier of this PlaceholderProvider instance.'
---

# <api__abstract></api__abstract> <api__class></api__class> PlaceholderProvider

Abstract class that is used to provide your own Placeholder patterns for AdvancedServerList to parse.

In order for your class to be considered a valid PlaceholderProvider will you need to set the `identifier` to a non-null, non-empty value without having any spaces in it.  
Once set, use [`AdvancedServerListAPI#addPlaceholderProvider(PlaceholderProvider)`](advancedserverlistapi.md#addplaceholderprovider(placeholderprovider)) to register your class for AdvancedServerList to use.
```

![grafik](https://github.com/mkdocs/mkdocs/assets/11576465/c33f3034-461c-4ba3-9b38-d2c9b6b504bd)

</p>
</details>

If for whatever reason you're interested about the structure of these options (And which ones I have) did I make a info page which isn't included in the built page, but can be found in the source here:
https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/.overrides/.schema.md

**Comment by pawamoy** [comment]
Nice work. Do I understand correctly that you are writing your API docs (as structured YAML) in the front matter directly? Is it duplicated from the source code?

I wonder: did you consider creating a Java handler for mkdocstrings :slightly_smiling_face:? If not, I'm still interested in *what* information you render here, and *how* you make it look :slightly_smiling_face: No additional input from you needed, I can look in details at what you show off here, but feel free to comment further or provide more info :smile:

**Reply by Andre601** [reply]
> Nice work. Do I understand correctly that you are writing your API docs (as structured YAML) in the front matter directly? Is it duplicated from the source code?

Yes you understand it right. The data is within the YAML frontmatter of the page.
You can find an example here: https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/api/reference/api/advancedserverlistapi.md?display=source#L1-L41

I use a custom template file to generate and render the content ([api-doc.html](https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/.overrides/api-doc.html)) which pulls its data from frontmatter through the `page.meta` jinja variable.

It's not directly *duplicated* if you mean copying it over. I wrote the Frontmatter completely by hand. Tho, I tried to copy over the same lines and to generate a structure that looks similar to actual Javadocs (Will explain this below).

> I wonder: did you consider creating a Java handler for mkdocstrings 🙂? If not, I'm still interested in _what_ information you render here, and _how_ you make it look 🙂 No additional input from you needed, I can look in details at what you show off here, but feel free to comment further or provide more info 😄

Yes, I considered it many... many times. But each time I got discouraged by the fact that while it may support any coding language through handlers, the core is still python, which I currently do not have the mental space for to learn...
If it was as simple as copying a template, tell it what parts to look for and then generate the stuff (i.e. with a template) would this be a different story. But from what I gathered is it not *that* simple and still requires Python programming language knowledge to some extend.

As pointed out do I use YAML frontmatter (I gave the overall schema for how it's structured in my initial post) to get the data to use.
I have one core template file, [`api-doc.html`](https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/.overrides/api-doc.html) which I use in a page. This template simply checks for specific frontmatter values being available to then include other files for those specific things.
As an example, if the `constructors` frontmatter is available (And not empty) will it include the [`constructor-summary.html`](https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/.overrides/api-objects/constructor-summary.html) and [`constructor-detail.html`](https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/.overrides/api-objects/constructor-detail.html) files for the "Constructor Summary" and "Constructor Detail" parts respectively. These will then simply iterate through the entries and generate the HTML stuff.

The structure of these docs is inspired by the actual Javadocs that can be rendered.

Also, if it is of any help will I quickly explain, to the best of my knowledge, how Javadocs are structured.

First of all, Javadocs are indicated by the first line starting with `/**` instead of the usual `/*` used by normal multiline comments.
Most javadocs have `*` in front of each line. Those are optional as only the start and end indicator (`*/`) are required here, but it's common knowledge to still have them for easier looking at docs.

Next are there "tags". These are indicated using a `@` symbol followed by the identifier (i.e. `@returns`).
Then there are also tags (I assume. Don't know what name those have) in the form of `{@indicator options}` which is used for stuff like making a link, adding inline-code, etc.

The main docs themself are using normal HTML formatting, meaning you use `<br>` for line breaks, `<a>` for links (Outside the docs. Internal ones should use `{@link <path> [displayname]}` instead), etc.

Here's a general structure of a Javadoc comment. Note that not all options would be available everywhere, the more detailed example further down should give a rough idea. Also, this is from my knowledge. Own research is recommended.
```java
/**
 * <description>
 * 
 * @author <name>
 * 
 * @version <version>
 * 
 * @param <name> <description>
 * 
 * @return <description>
 * 
 * @throws <path> <description>
 * 
 * @see <link|path> [displayname]
 * 
 * @since <version>
 * 
 * @deprecated <description>
 */
```

Any text that isn't started with a tag and is at the start is treated as the main comment. As mentioned before does it use normal HTML formatting here.

`@author` simply allows to define authors of this method, class, etc.

`@version` is similar to `@author` but allows to describe a version to display.

`@param` indicate a parameter that is part of the constructor, method/function, etc. The first word should be a matching parameter name while everything after is treated as the parameter's description.
It's worth noting that any tag doesn't care about leading/trailing spaces. It will trim everything accordingly. Also, descriptions can be on new lines if desired, which is what I do. As an example...
```
/**
 * @param something
 *        This is the param description
 */
```
...this is perfectly valid by the Javadoc generator.

`@return` is used to describe what is being returned by the method. This is only available on methods that actually return something other than a void.

`@throws` is to list possible exceptions that could be thrown and why. The first argument has to bee a path to the Exception class in question while everything else is treated as a description for it. Just like with `@param` can you put the description at a new line or add spaces to vertically align everything.

`@see` can be used to add links (Internal or external) to additional sources of information which would then be displayed in a "See also" section of the docs.

`@since` is used to indicate since what version a class, method/function, etc. exists.

`@deprecated` is used to add a Deprecation warning to the docs with an optional explanation as to why.

On to the `{@...}` ones, there is `{@link <path> [displayname]}` which will create a `<a ...><code>...</code></a>` object with the target name or the display name.
As an example `{@link com.example.lib.Example}` will render as `Example` while `{@link com.example.lib.Example Something}` renders as `Something`. Both would link to the class `Example` in `com.examble.lib`
You can also link to a specific section by adding `#methodName()` to the path (i.e. `{@link com.example.lib.Example#method(String)}`)
Something to note here is, that if the Java class itself already imports an object you want to link to through a `import` at the top, do you only need to provide the class name without any paths.

Next is there `{@code ...}` which for all I know is just a convenience method for wrapping stuff in `<code></code>` html-tags.
It can be used with `<pre>` html-tags for code blocks and content can be multiline, meaning that...
```
/**
 * <pre>{@code
 * {
 *     "something": "Here"
 * }
 * }</pre>
 */
```
...should render as a multi-line codeblock.

aaaand those are already the ones I know of, as those are the only I used so far...

I'll give a quick example to showcase the things in action. You can check [my api](https://github.com/Andre601/asl-api/tree/master) for more Javadocs which also use some different formatting.

How javadocs *could* look like when rendered can be seen here (Official Java 17 javadocs): https://docs.oracle.com/en/java/javase/17/docs/api/

```java
package com.example.lib;

/**
 * Classes can be described here. The comment would appear at the very top of the page to read.
 * 
 * @author  Andre_601 <github@andre601.ch>
 * @version v2.0.0
 */
public class Example {
    
    private final String foo;
    
    /**
     * Creates a new instance of this Example class.
     */
    public Example() {
        this.foo = "foo";
    }
    
    /**
     * Creates a new instance of this Example class with the provided String as content.
     *
     * @param str The String to add.
     * 
     * @see Example()
     */
    public Example(String str) {
        this.foo = str;
    }

    /**
     * This is a method that would return a String named {@code foo}.
     *
     * @return String named {@code foo}.
     */
    public String getFoo() {
        return "foo";
    }
    
    /**
     * This method checks if the provided String is {@code foo}.
     *
     * @param str The String to check
     *
     * @return True if the String is {@code foo} otherwise false.
     *
     * @since v2.0.0
     */
    public boolean isFoo(String str) {
        return str.equals("foo");
    }
    
    /**
     * A method that simply throws, because why not...
     * 
     * @throws java.lang.IllegalStateException Thrown when you call this method :P
     */
    public void throw() throws IllegalStateException {
        throw new IllegalStateException("You called this method dummy!");
    }
    
    /**
     * This method checks if the provided String is not {@code bar}.
     *
     * @param str The String to check
     *
     * @return True if the String is not {@code bar} otherwise false.
     *
     * @deprecated Replaced by {@link #isFoo(String)}
     */
    @Deprecated
    public boolean isNotBar(String str) {
        return !str.equals("bar");
    }
}
```

**Reply by Andre601** [reply]
I also found these sources that explain structures of Javadocs a little (Also first shows recommended order of the tags and structure):

- https://gist.github.com/ashishrana160796/4ac492caa9b306eac0adb3716e8b76d3
- https://binfalse.de/2015/10/05/javadoc-cheats-sheet/

**Reply by pawamoy** [reply]
Thank you very much for this detailed answer! I have experimented with such doc comments in the past, so I'm not a complete stranger to them :slightly_smiling_face: I had actually coded a Python package, [shellman](https://github.com/pawamoy/shellman) that reads similar comments from shell scripts, bringing docstrings to the shell!

Maybe I missed it in your answer, but I guess there are tools that are able to read these doc comments from Java sources and output them in a machine-readable format, for example JSON? Do you know any? It would be super helpful for developing a Java handler: just use the tool to collect the data, and write some templates to render it.

Also, yeah, developing handlers still demands writing a bit of Python (and also understanding how mkdocstrings and autorefs work, so reading more Python). I understand it can be off-putting for developers who are more proficient in other languages :slightly_smiling_face:

**Reply by Andre601** [reply]
> Maybe I missed it in your answer, but I guess there are tools that are able to read these doc comments from Java sources and output them in a machine-readable format, for example JSON? Do you know any? It would be super helpful for developing a Java handler: just use the tool to collect the data, and write some templates to render it.

I so far only know of the default Javadoc generator that generates HTML pages with the content from the comments... Not sure if there's one that creates JSON outputs here, but maybe google has the answer for us 😅

**Reply by Andre601** [reply]
Looking a bit into it, there seems to be some solutions to be available... For once [javadoc-json-doclet](https://github.com/RaidAndFade/javadoc-json-doclet) which creates JSON files from the Javadoc comments.

**Reply by Andre601** [reply]
Looking further at this, the above doclet no longer works as since Java 13 the `com.sun.javadoc` API support has been removed, which is bad given I myself run Java 17...
Not sure what a good alternative would be here now...

**Reply by pawamoy** [reply]
I'd myself go with the treesitter Python package, which provides parsers for many popular languages. Then I'd write an AST visitor, similar to what I do in Griffe, to extract information :shrug: Pretty involved though.

**Comment by Andre601** [comment]
Just want to mention 2 changes I made to this whole thing:

- I now generate the whole method, constructor, etc. using a combination of the name and provided parameter types.
  This means that something like this:
  ```yaml
  methods:
    - name: 'someMethod'
      parameters:
        - name: 'something'
          type: 'String'
        - name: 'somethingElse'
          type: 'int'
  ```
  will generate `someMethod(String, int)`
- I've updated and moved the schema file (It's now called structure). It's also included in the generated docs but not within the nav. Link is [here](https://docs.andre601.ch/AdvancedServerList/api/reference/structure/)

One thing I would love to do is to reduce duplicate stuff... As of right now am I re-using the same stuff over and over again...
Maybe with a macro this could be done... Tho at that point could a plugin or extension perhaps be the better alternative to consider...

**Reply by Andre601** [reply]
Aaaaand I did it... I made a macros to get rid of duplicate content. Would apreciate feedback from people with better jinja2 experience if there is ways to improve and further simplify it.

https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/.overrides/api-objects/info-template.html
